### PR TITLE
fix windows test_senet CI

### PR DESCRIPTION
**Pull request by rijobro**
### Description
Fix windows test_senet by not reusing networks for different tests.

Also speed up all uses of `test_script_save`, which will be performed with CUDA if available, falling back to the CPU if not.

### Status
**Ready**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [x] Integration tests passed locally by running `./runtests.sh --codeformat --coverage`.
- [x] Quick tests passed locally by running `./runtests.sh --quick`.
- [x] In-line docstrings updated.
- [x] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

**Comment by wyli** [comment]
thanks for looking into this! I couldn't figure out the root cause for the windows test, was about to submit a PR to remove the test as well...

**Comment by rijobro** [comment]
@wyli It seems that the windows test was still failing in my initial commit so i've commented out the pre-trained test to see if that fixes. If not, I'll comment out the other tests until I figure out which is the problematic one.

**Comment by rijobro** [comment]
@wyli @ericspod that seems to have fixed it, the problem was reusing the network for two different tests (`test_senet_shape` and `test_script`).

**Comment by wyli** [comment]
> @wyli @ericspod that seems to have fixed it, the problem was reusing the network for two different tests (`test_senet_shape` and `test_script`).

thanks, there's another issue now with `torch.allclose` for the cuda tensors. I've checked the default `atol=1e-8` which is too strict... `torch.allclose(out_orig, out_reloaded, atol=1e-7)` works fine for me locally

**Comment by rijobro** [comment]
@wyli done. Moved the comparison inside of the `test_script_save` function.

**Comment by wyli** [comment]
ok, looks like 1e-7 is still too small... still need to increase atol

File: tests/utils.py [hunk 1]
```diff
@@ -143,7 +151,15 @@ def test_script_save(net, *inputs, eval_nets=True):
         result1 = net(*inputs)
         result2 = reloaded_net(*inputs)
```
Comment (wyli) [inline tests/utils.py:153]: not 100% sure if it's relevant, perhaps we need to set the deterministic flag to ensure the reproducibility...
https://pytorch.org/docs/stable/notes/randomness.html
```suggestion
        monai.utils.set_determinism(seed=0)
        result1 = net(*inputs)
        result2 = reloaded_net(*inputs)
        monai.utils.set_determinism(seed=None)
```

**Comment by ericspod** [comment]
I hadn't every encountered the precision issue with `allclose`, either in Torch or Numpy as you have it now. Is this a precision issue with GPU computation?

**Comment by wyli** [comment]
> I hadn't every encountered the precision issue with `allclose`, either in Torch or Numpy as you have it now. Is this a precision issue with GPU computation?

yes but it works fine with pytorch 1.5

**Comment by rijobro** [comment]
I'm getting some pretty wild differences now (e.g., [here](https://github.com/Project-MONAI/MONAI/pull/1268/checks?check_run_id=1426105557#step:6:7741)), so that might indicate a bug. I'll switch the computation back to the CPU tomorrow to see if it's my fault or not.

P.S. I switched from `torch.allclose` to `np.testing.all_close` as this prints the relative and absolute differences, which is obviously helpful during debugging.

File: tests/test_senet.py [hunk 2]
```diff
@@ -26,31 +26,33 @@
...
-TEST_CASE_4 = [se_resnet152(3, 2, 2)]
-TEST_CASE_5 = [se_resnext50_32x4d(3, 2, 2)]
-TEST_CASE_6 = [se_resnext101_32x4d(3, 2, 2)]
+net_args = {"spatial_dims": 3, "in_channels": 2, "num_classes": 2}
+TEST_CASE_1 = [senet154, net_args]
+TEST_CASE_2 = [se_resnet50, net_args]
+TEST_CASE_3 = [se_resnet101, net_args]
+TEST_CASE_4 = [se_resnet152, net_args]
+TEST_CASE_5 = [se_resnext50_32x4d, net_args]
+TEST_CASE_6 = [se_resnext101_32x4d, net_args]
```
Comment (YuanTingHsieh) [inline tests/test_senet.py:35]: let's call this NET_ARGS then since it is a constant kind of?
and to avoid warnings below (cuz we have net_args below as well)

File: tests/utils.py [hunk 3]
```diff
@@ -123,27 +123,44 @@ def setUp(self):
...
-
-    return result1, result2
+        set_determinism(seed=None)
+    # When using e.g., VAR, we will produce a tuple of outputs.
+    # Hence, convert all to tuples and then compare all elements.
+    if not isinstance(result1, tuple):
+        retult1 = (result1,)
+        retult2 = (result2,)
```
Comment (YuanTingHsieh) [inline tests/utils.py:159]: typos here
should be 
result1 = (result1, )
result2 = (result2, )

File: tests/utils.py [hunk 4]
```diff
@@ -123,27 +123,44 @@ def setUp(self):
...
+
+    for i, j in zip(result1, result2):
+        if None not in (i, j):  # might be None
+            np.testing.assert_allclose(i.cpu(), j.cpu())
```
Comment (YuanTingHsieh) [inline tests/utils.py:163]: when I modify this line to
`np.testing.assert_allclose(i.detach().cpu().numpy(), j.detach().cpu().numpy(), atol=1e-5)`
tests are passed for me

Comment (YuanTingHsieh) [reply tests/utils.py:163]: something like this: https://github.com/YuanTingHsieh/MONAI/commit/81a337428878c6114d3698ad0e2d7907187d95e3

**Comment by rijobro** [comment]
Thanks @YuanTingHsieh!

**Comment by rijobro** [comment]
Worryingly, using the GPU at all requires a relative tolerance of > 0.1. Does this imply that if our network is already on the GPU we can't safely save it and get reliable results? Only for certain OS's it seems.

**Comment by wyli** [comment]
> Worryingly, using the GPU at all requires a relative tolerance of > 0.1. Does this imply that if our network is already on the GPU we can't safely save it and get reliable results when we load it back in? Only for certain OS's it seems.

for a workaround for now, shall we roll back the CPU based torchscript tests, and file a new ticket for the GPU based issue?
currently the windows ci is broken we couldn't check any new PR properly

**Comment by rijobro** [comment]
Sounds good to me, i'll update now. Will have to play around with the `rtol` and `atol` quickly to find reasonable values for the CPU.

**Comment by rijobro** [comment]
all done and GPU issue created: #1273.

**Review by wyli — APPROVED** [review]
